src/Psl/Type/Internal/MutableSetType.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use Psl\Collection;
+use Psl\Str;
+use Psl\Type;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Throwable;
+
+use function is_iterable;
+use function is_object;
+
+/**
+ * @template T of array-key
+ *
+ * @extends Type\Type<Collection\MutableSetInterface<T>>
+ *
+ * @internal
+ */
+final readonly class MutableSetType extends Type\Type
+{
+    /**
+     * @psalm-mutation-free
+     *
+     * @param Type\TypeInterface<T> $key_type
+     */
+    public function __construct(
+        private readonly Type\TypeInterface $key_type,
+    ) {
+    }
+
+    /**
+     * @throws CoercionException
+     *
+     * @return Collection\MutableSetInterface<T>
+     */
+    public function coerce(mixed $value): Collection\MutableSetInterface
+    {
+        if (is_iterable($value)) {
+            /** @var Type\Type<T> $key_type */
+            $key_type = $this->key_type;
+            /** @var list<T> $keys */
+            $keys = [];
+            $key = null;
+            $iterating = true;
+            try {
+                /**
+                 * @var T $key
+                 */
+                foreach ($value as $key => $_) {
+                    $iterating = false;
+                    $keys[] = $key_type->coerce($key);
+                    $iterating = true;
+                }
+            } catch (Throwable $e) {
+                throw match (true) {
+                    $iterating => CoercionException::withValue(null, $this->toString(), PathExpression::iteratorError($key), $e),
+                    default => CoercionException::withValue($key, $this->toString(), PathExpression::iteratorKey($key), $e),
+                };
+            }
+
+            /** @var Collection\MutableSet<T> */
+            return new Collection\MutableSet($keys);
+        }
+
+        throw CoercionException::withValue($value, $this->toString());
+    }
+
+    /**
+     * @throws AssertException
+     *
+     * @return Collection\MutableSetInterface<T>
+     *
+     * @psalm-assert Collection\MutableSetInterface<T> $value
+     */
+    public function assert(mixed $value): Collection\MutableSetInterface
+    {
+        if (is_object($value) && $value instanceof Collection\MutableSetInterface) {
+            /** @var Type\Type<T> $key_type */
+            $key_type = $this->key_type;
+            /** @var list<T> $keys */
+            $keys = [];
+            $key = null;
+            $iterating = true;
+            try {
+                /**
+                 * @var T $key
+                 */
+                foreach ($value as $key => $_) {
```

Comment:
also here it might make more sense to assert the values instead of the key.
(assuming we'dd allow something broader than array-key to be contained in the set)